Python/Product/VSInterpreters/CPythonInterpreterFactoryProvider.cs

```diff
@@ -230,77 +221,90 @@ class CPythonInterpreterFactoryProvider : IPythonInterpreterFactoryProvider {
-                    using (var python = baseKey.OpenSubKey(PythonPath)) {
-                        if (python != null) {
-                            anyChanged |= RegisterInterpreters(registeredPaths, registeredIds, true, python, arch);
-                        }
-                    }
-                }
+            HashSet<string> registeredPaths = new HashSet<string>();
+            var arch = Environment.Is64BitOperatingSystem ? null : (ProcessorArchitecture?)ProcessorArchitecture.X86;
 
-                using (var baseKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry32))
+            Dictionary<string, InterpreterInformation> userFactories = new Dictionary<string, InterpreterInformation>();
+            using (var baseKey = RegistryKey.OpenBaseKey(RegistryHive.CurrentUser, RegistryView.Default)) {
                 using (var python = baseKey.OpenSubKey(PythonPath)) {
                     if (python != null) {
-                        anyChanged |= RegisterInterpreters(registeredPaths, registeredIds, false, python, ProcessorArchitecture.X86);
+                        RegisterInterpreters(userFactories, registeredPaths, true, python, arch);
                     }
                 }
+            }
 
-                if (Environment.Is64BitOperatingSystem) {
-                    using (var baseKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry64))
-                    using (var python64 = baseKey.OpenSubKey(PythonPath)) {
-                        if (python64 != null) {
-                            anyChanged |= RegisterInterpreters(registeredPaths, registeredIds, false, python64, ProcessorArchitecture.Amd64);
-                        }
+            Dictionary<string, InterpreterInformation> machineFactories = new Dictionary<string, InterpreterInformation>();
+            using (var baseKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry32))
+            using (var python = baseKey.OpenSubKey(PythonPath)) {
+                if (python != null) {
+                    RegisterInterpreters(machineFactories, registeredPaths, false, python, ProcessorArchitecture.X86);
+                }
+            }
+
+            if (Environment.Is64BitOperatingSystem) {
+                using (var baseKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry64))
+                using (var python64 = baseKey.OpenSubKey(PythonPath)) {
+                    if (python64 != null) {
+                        RegisterInterpreters(machineFactories, registeredPaths, false, python64, ProcessorArchitecture.Amd64);
+                    }
+                }
+            }
+
+            // Then update our cached state with the lock held.
+            lock (this) {
+                var uniqueIds = userFactories.Select(x => x.Key)
+                       .Concat(machineFactories.Select(x => x.Key))
+                       .Distinct();
```

Comment:
Put into a set to avoid multiple evaluations/breakage (if we later change userFactories or machineFactories in the first for loop)

You could also use `.Keys` instead of the `.Select`, right? And `.Union()` instead of `.Select().Distinct()`. Or maybe I'm missing something there.